lib/rucio/core/replica.py

```diff
@@ -864,13 +921,18 @@ def get_multi_cache_prefix(cache_site, filename, logger=logging.log):
 
 
 def _list_replicas(dataset_clause, file_clause, state_clause, show_pfns,
-                   schemes, files, rse_clause, rse_expression, client_location, domain,
-                   sign_urls, signature_lifetime, constituents, resolve_parents,
+                   schemes, files_wo_replica, rse_clause, client_location, domain,
+                   sign_urls, signature_lifetime, constituent_clause, resolve_parents,
                    updated_after, filters, ignore_availability,
                    session):
 
-    files = [dataset_clause and _list_replicas_for_datasets(dataset_clause, state_clause, rse_clause, ignore_availability, updated_after, session),
-             file_clause and _list_replicas_for_files(file_clause, state_clause, files, rse_clause, ignore_availability, updated_after, session)]
+    # iterator which merges multiple sorted replica sources into a combine sorted result without loading everything into the memory
+    replicas = heapq.merge(
+        _list_replicas_for_datasets(dataset_clause, state_clause, rse_clause, ignore_availability, updated_after, session),
+        _list_replicas_for_files(file_clause, state_clause, files_wo_replica, rse_clause, ignore_availability, updated_after, session),
+        _list_replicas_for_constituents(constituent_clause, state_clause, files_wo_replica, rse_clause, ignore_availability, updated_after, session),
```

Comment:
this looks nice to me.

maybe worth to add some comment that `_files` and `_constituent` are very similar, and if one of the SQLAlchemy queries changes in one, that the other must be adapted as well? Otherwise we run the risk of having two queries with different outcomes.

Reply:
It's even more than that: all 3 of these functions are very similar. There is an initial query, which is different, but then we append filters to it, which are all identical. The `_files`  has a slightly different structure, but I think it's only for historical reasons, not a technical one. It's maybe worth to extract common code (appending filters to initial query) into a separate function ? WDYT ?

Reply:
yeah, should be done. but can be a separate issue/PR